finite_approx/gibbs_sampler_lib.py

```diff
@@ -101,3 +103,213 @@ def sample(self, burnin, num_gibbs_draws):
+        self.sigma_a = sigma_a
+
+        self.initialize_sampler()
+        self.lh_quantities()
+
+    def initialize_sampler(self):
+        # Initial values for draws
+        self.z = np.random.binomial(1, 0.5, [self.x_n, self.k_approx ])
+        self.z = self.z.astype(float)
+        self.z2 = np.dot(self.z.T, self.z)
+
+        self.z_draws = []
+
+    def lh_quantities(self): # intermediate quantities to compute likelihoods
+        self.var = np.dot(self.z.T, self.z)\
+                + self.sigma_eps/self.sigma_a * np.eye(self.k_approx)
+        self.var_inv = np.linalg.inv(self.var)
+
+        [_, logconst] = np.linalg.slogdet(self.var)
+        mean_a = np.dot(np.linalg.solve(self.var, self.z.T), self.x)
+        self.log_lh_X = -1/(2*self.sigma_eps) * \
+                np.trace(np.dot(self.x.T, self.x - np.dot(self.z, mean_a)) )\
+                - (self.x_d/2) * logconst
+
+    def draw_z(self, keep_draw=False):
+        for n in range(self.x_n):
+            for k in range(self.k_approx):
+                # first, compute p(z_nk = 1 | Z_{-nk})
+                # This is equation (6) in Griffiths and Ghahramani
+                # http://mlg.eng.cam.ac.uk/zoubin/papers/ibp-nips05.pdf
+                p_z_nk1_cond_Z = (np.sum(self.z[:, k]) - self.z[n, k] + \
+                                self.alpha/self.k_approx)\
+                                /(self.x_n + self.alpha/self.k_approx)
+                assert (p_z_nk1_cond_Z >= 0) & (p_z_nk1_cond_Z <= 1)
+
+                if self.z[n,k] == 1:
+                    p_znk_cond_z_stay = p_z_nk1_cond_Z
+                    p_znk_cond_z_update = 1 - p_z_nk1_cond_Z
+                else:
+                    p_znk_cond_z_update = p_z_nk1_cond_Z
+                    p_znk_cond_z_stay = 1 - p_z_nk1_cond_Z
+
+                # next compute p(X|Z_new)
+                # equation (8) in Griffiths and Ghahramani
+
+                # flip z[n,k]
+                self.z[n,k] = 1 - self.z[n,k]
+                [log_lh_X_update, var_inv_update] = update_x_lp_cond_z(\
+                    self.x, self.x2, self.z2, self.z, self.var_inv, self.sigma_eps, \
+                    self.sigma_a, self.k_approx, n, k)
+
+                log_p_znk_cond_z_update = np.log(p_znk_cond_z_update)
+                log_p_zn_update = log_lh_X_update + log_p_znk_cond_z_update \
+                    - sp.misc.logsumexp([log_lh_X_update + log_p_znk_cond_z_update,\
+                        self.log_lh_X+ np.log(p_znk_cond_z_stay)])
+
+                #assert (np.exp(log_p_zn_update) >= 0) & (np.exp(log_p_zn_update) <= 1)
+
+                choice = np.random.binomial(1, np.exp(log_p_zn_update))
+
+                # update Z and prep log likelihood for next round:
+                if choice == 1: # update the necessary quantities
+                    self.log_lh_X = deepcopy(log_lh_X_update)
+                    self.var_inv = deepcopy(var_inv_update)
+                    self.z2 = np.dot(self.z.T, self.z)
```

Comment:
You have only changed one term -- no need to re-calculate the dot product of the whole big matrix, which is very expensive.